source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+            SpvStorageClassFunction)
+      break;
+  }
+  return (type_itr != module_->types_values_end()) ? type_itr->result_id() : 0;
+}
+
+uint32_t InlinePass::AddPointerToType(uint32_t typeId) {
+  uint32_t resultId = TakeNextId();
+  std::vector<ir::Operand> in_operands;
+  in_operands.emplace_back(
+      spv_operand_type_t::SPV_OPERAND_TYPE_STORAGE_CLASS,
+      std::initializer_list<uint32_t>{uint32_t(SpvStorageClassFunction)});
+  in_operands.emplace_back(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                           std::initializer_list<uint32_t>{uint32_t(typeId)});
+  std::unique_ptr<ir::Instruction> type_inst(
+      new ir::Instruction(SpvOpTypePointer, 0, resultId, in_operands));
+  module_->AddType(std::move(type_inst));
+  return resultId;
+}
+
+void InlinePass::AddBranch(uint32_t labelId,
+                           std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> branch_in_operands;
+  branch_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{labelId}));
+  std::unique_ptr<ir::Instruction> newBranch(
+      new ir::Instruction(SpvOpBranch, 0, 0, branch_in_operands));
+  (*bp)->AddInstruction(std::move(newBranch));
+}
+
+void InlinePass::AddStore(uint32_t ptrId, uint32_t valId,
+                          std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> store_in_operands;
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{valId}));
+  std::unique_ptr<ir::Instruction> newStore(
+      new ir::Instruction(SpvOpStore, 0, 0, store_in_operands));
+  (*bp)->AddInstruction(std::move(newStore));
+}
+
+void InlinePass::AddLoad(uint32_t typeId, uint32_t resultId, uint32_t ptrId,
+                         std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> load_in_operands;
+  load_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  std::unique_ptr<ir::Instruction> newLoad(
+      new ir::Instruction(SpvOpLoad, typeId, resultId, load_in_operands));
+  (*bp)->AddInstruction(std::move(newLoad));
+}
+
+// Generate callee code into newBlocks to be inlined for the function call at
+// call_ii. Also add new function variables into caller func
+
+void InlinePass::GenInlineCode(
+    std::vector<std::unique_ptr<ir::BasicBlock>>* newBlocks,
+    std::vector<std::unique_ptr<ir::Instruction>>* newVars,
+    ir::UptrVectorIterator<ir::Instruction> call_ii,
+    ir::UptrVectorIterator<ir::BasicBlock> call_bi) {
+  // Map from callee id to caller id
+  std::unordered_map<uint32_t, uint32_t> callee2caller;
+  // Pre-call OpSampledImage Insts
```

Comment:
Are you open to a recursive algorithm here? Essentially, I think I can implement this by only adding a recursive call to CloneSameBlockOps just after cloning the same block instruction, which would generate any same-block-ops needed for the cloned instruction. Otherwise I am going to split this algorithm into two loops, one to build up the stack of instructions to insert, and one to walk back down it.